### PR TITLE
Add `then` method for composing two Clifford tableaux

### PR DESCRIPTION
**Pull request by ybc1991**
The preparation step for making general Clifford Gates #3639.

The basic logic of implementation is tracking the transformation of all generators of Pauli group. Namely, list the transform of first tableau:`U_1X_iU_1' = r_i \Prod_j X_j^{t_(i,j)} Z_i^{s_(i,j)}` and list the same thing for second tableau `U_2`, then combining them.

With this method, it will allow us to build other unnamed Clifford Gates by composition.

### CODE REVIEW

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 1]
```diff
@@ -84,6 +83,9 @@ def rs(self, new_rs: np.array) -> None:
         assert np.shape(new_rs) == (2 * self.n,)
         self._rs[:-1] = np.array(new_rs).astype(bool)
 
+    def matrix(self) -> np.array:
```
Comment (mpharrigan) [inline cirq-core/cirq/qis/clifford_tableau.py:86]: docstring

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 2]
```diff
@@ -176,6 +178,43 @@ def _str_full_(self) -> str:
 
         return string
 
+    def merged_with(self, second: 'CliffordTableau') -> 'CliffordTableau':
+        """Returns a merged CliffordTableau of self tableau and second tableau.
+
+        The meaning of merge two clifford tableau is the corresponding unitary operation
```
Comment (mpharrigan) [inline cirq-core/cirq/qis/clifford_tableau.py:184]: Nit: we should pick a plural. If we're anglecizing: "tableaus", keeping the french origins "tableaux". Meriam Webster has both as acceptable.

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 3]
```diff
@@ -176,6 +178,43 @@ def _str_full_(self) -> str:
...
+        of merged clifford tableau is equal to (up to global phase) the composed unitary
+        operation of self tableau and the one of second tableau.
```
Comment (mpharrigan) [inline cirq-core/cirq/qis/clifford_tableau.py:186]: This description is a little clunky, and I think it's essential to be very clear about what this does. "Merge" is a very ambiguous term. In fact: maybe this should be called `compose_with`?

"A merged [/composed] tableau is equal to (up to global phase) the composed unitary operation of the two tableaux"

Also say which one happens first, since I don't really know. Instead of "self", maybe say "this tableau" since it will be called `tableau.merged_with(tab2)`

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 4]
```diff
@@ -176,6 +178,43 @@ def _str_full_(self) -> str:
...
+        """
+        m1 = self.matrix().astype(int)
+        m2 = second.matrix().astype(int)
```
Comment (mpharrigan) [inline cirq-core/cirq/qis/clifford_tableau.py:189]: why is the astype needed? shouldn't we keep these as the correct type in this class?

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 5]
```diff
@@ -176,6 +178,43 @@ def _str_full_(self) -> str:
...
+
+        merged_m = np.mod(m1.dot(m2), 2)
+        phase = np.mod(p1 + m1.dot(p2), 2)
+
+        # we need more phase correction for expanding Y to XZ and swapping Z_iX_i order.
+        for k in range(2 * self.n):
+            swap_phase = 0  # value betwen 0 and 3 representing [1, i, -1, -i] respectively.
+            prev_row_sum = np.zeros([2 * self.n])
+            swap_phase += np.sum(m1[k, : self.n] * m1[k, self.n :])  # Y gate => iXZ
+            for i, v in enumerate(m1[k]):
```
Comment (mpharrigan) [inline cirq-core/cirq/qis/clifford_tableau.py:202]: Can this be done with a vectorized numpy operation? My "slow python iteration over numpy arrays" senses are tingling.

If you need to only do it where the entries are not equal to zero, you can do fancy indexing:

```python
non_zero = m1[k] != 0
swap_phase += np.sum(m2[non_zero, :, self.n] * m2, [non_zero] ....
```

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 6]
```diff
@@ -176,6 +178,43 @@ def _str_full_(self) -> str:
...
+                if v == 0:
+                    continue
+                swap_phase += np.sum(m2[i, : self.n] * m2[i, self.n :])  # Y gate => iXZ
+                swap_phase += 2 * np.sum(m2[i, : self.n] * prev_row_sum[self.n :])  # swap Z_i X_i
```
Comment (mpharrigan) [inline cirq-core/cirq/qis/clifford_tableau.py:206]: weird formatting with the space between the `:` and `self.n`, here and elsewhere. Normally slicing looks like `list[:i]`

**Review by mpharrigan — CHANGES_REQUESTED** [review]
I can't speak to code correctness, I assume you've gotten that part right. Added some style points. Could you also expand on this PRs description: how will this contribute towards the linked issue?

File: cirq-core/cirq/qis/clifford_tableau_test.py [hunk 7]
```diff
@@ -322,3 +322,91 @@ def test_deprecated_clifford_location():
...
+    _ = [_S(t, 0) for t in (t2, expected_t)]
+    assert expected_t == t1.merged_with(t2)
+
+    t1, t2, expected_t = three_identical_table(1)
+    _ = [_X(t, 0) for t in (t1, expected_t)]
+    _ = [_H(t, 0) for t in (t1, expected_t)]
+    _ = [_Z(t, 0) for t in (t1, expected_t)]
+    _ = [_S(t, 0) for t in (t2, expected_t)]
+    _ = [_H(t, 0) for t in (t2, expected_t)]
+    assert expected_t == t1.merged_with(t2)
```
Comment (smitsanghavi) [inline cirq-core/cirq/qis/clifford_tableau_test.py:353]: It's worth adding a test case that tests `expected_t != t1.merged_with(t2)` when `t2` is acting before `t1` i.e. Enforce the order of composition.

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 8]
```diff
@@ -176,6 +178,43 @@ def _str_full_(self) -> str:
 
         return string
 
+    def merged_with(self, second: 'CliffordTableau') -> 'CliffordTableau':
```
Comment (smitsanghavi) [inline cirq-core/cirq/qis/clifford_tableau.py:181]: How about making this `__mul__` (or `__rmul__` if you want to reverse the order)? It might be nice to be able to use `*` to multiply two tableaus

**Review by smitsanghavi — CHANGES_REQUESTED** [review]

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 8 above)
Comment (daxfohl) [reply cirq-core/cirq/qis/clifford_tableau.py:181]: I have another PR that merges (and unmerges) states. We should be consistent with names.

Comment (mpharrigan) [reply cirq-core/cirq/qis/clifford_tableau.py:181]: In that case, are you doing the tensor product? If yes, we shouldn't use "merge" for either!

Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py:181]: Let me give more background information.  The reason I choose `merge_with` is simple -- because it is used in `SingleCliffordGate`, see [here](https://github.com/quantumlib/Cirq/blob/master/cirq-core/cirq/ops/clifford_gate.py#L285). The meaning of this method is simply a conjugate mapping of two functions `U_2U_1 g U_1'U_2'`. (Or equivalently, just composed function if viewing in state vector;) I think it is different from the meaning of merge_state as @daxfohl mentioned.

As for @smitsanghavi  mentioned `__mul__` or other magic functions, the benefits of doing this are less in here. Because this is the helper function, I plan to use that in the CliffordGate. So personally, I prefer the `compose_with` so far. Anyway, we can discuss that in cirq-cync meeting?

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 5 above)
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py:202]: Unfortunately, that is not doable because of this line `prev_row_sum += m2[i]`. It is a stateful for-loop. If we indeed worry about the speed, we can consider Cython (NumPy is dependent on it anyway).

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 6 above)
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py:206]: I know. I feel that is weird as well. But that is how flynt/black formated. I have tried to remove the space and they added it back after formating.

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 4 above)
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py:189]: Hmm, the under Clifford Tableau is always 0, 1. So the underlying datatype is `bool`. (It has been that for long time). Boolean type computation is awkward in numpy, hence it is usually easier to transform to int type and do the mod 2 at last.

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 2 above)
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py:184]: Do you have a preference?  I don't mind any of them.

File: cirq-core/cirq/qis/clifford_tableau_test.py (hunk 7 above)
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau_test.py:353]: Yes, that sounds good

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 1 above)
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py:86]: done.

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 3 above)
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py]: Updated the comments. Also, use the `then` as the function name as discussed in cyqn meeting.

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 6 above)
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py:207]: So just leave it as what auto-format did.

**Comment by ybc1991** [comment]
> I can't speak to code correctness, I assume you've gotten that part right. Added some style points. Could you also expand on this PRs description: how will this contribute towards the linked issue?

Add more context in the description. Addressed the most comment and rename the function according to the Cyqn meeting votes. PTAL, thanks.

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 9]
```diff
@@ -178,12 +179,12 @@ def _str_full_(self) -> str:
...
-        """Returns a merged CliffordTableau of self tableau and second tableau.
+    def then(self, second: 'CliffordTableau') -> 'CliffordTableau':
+        """Returns a composed CliffordTableau of this tableau and the second tableau.
 
-        The meaning of merge two clifford tableau is the corresponding unitary operation
-        of merged clifford tableau is equal to (up to global phase) the composed unitary
-        operation of self tableau and the one of second tableau.
+        Then composed tableau is equal to (up to global phase) the composed
+        unitary operation of the two tableaux, i.e. equivalent to applying the unitary operation
+        this CliffordTableau then applying the second one.
```
Comment (mpharrigan) [inline cirq-core/cirq/qis/clifford_tableau.py:187]: nit: applying the unitary operation **of** this CliffordTableau

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 4 above)
Comment (mpharrigan) [reply cirq-core/cirq/qis/clifford_tableau.py:190]: maybe add a comment?

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 5 above)
Comment (mpharrigan) [reply cirq-core/cirq/qis/clifford_tableau.py:203]: thanks

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 6 above)
Comment (mpharrigan) [reply cirq-core/cirq/qis/clifford_tableau.py:207]: sorry for the confusion

**Review by mpharrigan — APPROVED** [review]

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 8 above)
Comment (mpharrigan) [reply cirq-core/cirq/qis/clifford_tableau.py]: for the record: we discussed in cirq-cync; had many votes; and settled on `then` as the method name and supporting `@` for matrix-multiply-like

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 4 above)
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py:191]: Done

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 9 above)
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py]: Done.

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 2 above)
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py]: Pick tableaux

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 10]
```diff
@@ -176,6 +179,49 @@ def _str_full_(self) -> str:
 
         return string
 
+    def then(self, second: 'CliffordTableau') -> 'CliffordTableau':
```
Comment (balopat) [inline cirq-core/cirq/qis/clifford_tableau.py:182]: Shouldn't we do some kind of dimensionality checking? I haven't see the cases handled/tested when the sizes of the tableaus are different.

**Review by balopat — CHANGES_REQUESTED** [review]
This looks good, I still have to parse the phase correction logic.

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 11]
```diff
@@ -176,6 +179,49 @@ def _str_full_(self) -> str:
...
+        """Returns a composed CliffordTableau of this tableau and the second tableau.
+
+        Then composed tableau is equal to (up to global phase) the composed
+        unitary operation of the two tableaux, i.e. equivalent to applying the unitary
+        operation of this CliffordTableau then applying the second one.
```
[the remark this answers is not in the document]
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py:187]: Done.

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 10 above)
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py:182]: Thanks for catching that. (If there is no check in the beginning, it will throw an error at the matrix multiplication step. But it may be confusing.) Add the corresponding test as well.

**Comment by ybc1991** [comment]
> This looks good, I still have to parse the phase correction logic.

The steps were modified by myself. (Different from Qiskit did by killing one-for loop) So Let me add an example here for the record about the phase correction. Each step in the code are corresponding to following cases:
image

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 4 above)
Comment (smitsanghavi) [reply cirq-core/cirq/qis/clifford_tableau.py:209]: Just for my understanding, what is awkward about the boolean computation in numpy in this context?
From a cursory look at the logic below, we seem to be doing a bunch of `+` operations followed by `np.mod(_, 2)`. Wouldn't just using the `^` on the bool numpy arrays be more readable (and maybe also efficient)?

**Review by smitsanghavi — APPROVED** [review]
Looks good to me. Added a minor comment regarding the array types.

**Comment by ybc1991** [comment]
> Looks good to me. Added a minor comment regarding the array types.

Hi Smit, I have tried the NumPy matrix multiplication with bool type. I don't think it is equivalent to matrix multiplication with int type then do the modulo 2. Try this simple example:
```python
t = np.array([[1, 0], [1,1]], dtype=bool)
z = np.array([[1, 1], [1,0]], dtype=bool)
print((t @ z).astype(int))
print(np.mod(t.astype(int) @ z.astype(int), 2))
```
The first one is an all 1 matrix and the second one is not.

**Comment by smitsanghavi** [comment]
> The first one is an all 1 matrix and the second one is not.

Thanks! That makes sense. The `bool` one is doing `OR` instead of `+` in the matrix multiplication.

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 12]
```diff
@@ -176,6 +179,67 @@ def _str_full_(self) -> str:
...
+            swap_phase += np.sum(m1[k, : self.n] * m1[k, self.n :])  # Y gate => iXZ
+            for i, v in enumerate(m1[k]):
+                if v == 0:
+                    continue
+                swap_phase += np.sum(m2[i, : self.n] * m2[i, self.n :])  # Y gate => iXZ
+                swap_phase += 2 * np.sum(m2[i, : self.n] * prev_row_sum[self.n :])  # swap Z_i X_i
+                prev_row_sum += m2[i]
+            prev_row_sum = np.mod(prev_row_sum, 2)
+            swap_phase -= np.sum(prev_row_sum[: self.n] * prev_row_sum[self.n :])  # XZ => -iY
+            phase[k] = (phase[k] + (swap_phase % 4) / 2) % 2
```
Comment (balopat) [inline cirq-core/cirq/qis/clifford_tableau.py:230]: Can you add more explanatory comments to these lines? A reference would be great too. It is really not obvious to me that why we are adding observables (rows) together for example to count the Ys..

**Review by balopat — CHANGES_REQUESTED** [review]

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 12 above)
Comment (balopat) [reply cirq-core/cirq/qis/clifford_tableau.py:230]: I get the %4, counting 1j phases for each ZX and the 1j * 1j for -1 for XZ. But the way we combine the second tableaux is just hard to parse from the code itself, which I think should be a criteria for readable code, or at least we should have a ref, so that someone can understand it from there. The original Aaronson and Gottesman paper did not talk about combining the tableaux.

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 13]
```diff
@@ -176,6 +179,67 @@ def _str_full_(self) -> str:
...
+
+        # we need more phase correction for expanding Y to XZ and swapping Z_iX_i order.
+        for k in range(2 * self.n):
+            swap_phase = 0  # value betwen 0 and 3 representing [1, i, -1, -i] respectively.
+            prev_row_sum = np.zeros([2 * self.n])
+            swap_phase += np.sum(m1[k, : self.n] * m1[k, self.n :])  # Y gate => iXZ
+            for i, v in enumerate(m1[k]):
+                if v == 0:
+                    continue
+                swap_phase += np.sum(m2[i, : self.n] * m2[i, self.n :])  # Y gate => iXZ
```
[the remark this answers is not in the document]
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py:225]: Well, it cannot. Note `if v==0: continue`, not every row will count.

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 14]
```diff
@@ -176,6 +179,67 @@ def _str_full_(self) -> str:
...
+
+        p1 = self.rs.astype(int)
+        p2 = second.rs.astype(int)
+
+        merged_m = np.mod(m1.dot(m2), 2)
+        phase = np.mod(p1 + m1.dot(p2), 2)
+
+        # we need more phase correction for expanding Y to XZ and swapping Z_iX_i order.
+        for k in range(2 * self.n):
+            swap_phase = 0  # value betwen 0 and 3 representing [1, i, -1, -i] respectively.
```
[the remark this answers is not in the document]
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py]: You are right. I should say this value modulo 4 representing [1, i, -1, -i] respectively.

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 12 above)
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py:231]: Sure, add more sentences. But it is really hard for me to describe it precisely. The key idea is global phase is the only scalar so that we can always move that to the beginning. So we can safely compute the phase of each component whenever we like then add them at last. We have lots of components to generate phases `(phase[k] + (swap_phase % 4) / 2) % 2` the first `phase[k]` is the phase we generate by composing the two tableaux without adjusting the order etc(note like origin table may have +/- phase for each stabilizer). 

I don't know a good reference for it. @Strilanc Do you know any good references for it?

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (balopat) [reply cirq-core/cirq/qis/clifford_tableau.py:225]: Sorry, I meant that we could do this before the outer loop:

```
        num_ys1 = [np.sum(m1[k, : self.n] * m1[k, self.n :]) for k in range(2 * self.n)]
        num_ys2 = [np.sum(m2[k, : self.n] * m2[k, self.n:]) for k in range(2 * self.n)]
```

And then here we could refer to that "cache": 

```
      swap_phase += num_ys2[i]  # Y gate => iXZ
```

[the remark this answers is not in the document]
Comment (balopat) [reply cirq-core/cirq/qis/clifford_tableau.py:225]: Not sure how much it saves - it depends on the number of columns that have 1 in the tableau...and for those who have all 0 columns (would that be a valid symplectic matrix? e.g. [[+IX +IX], [+IZ +IZ]]?), it would be more expensive...

[the remark this answers is not in the document]
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py:229]: I see your point now. It can be further vectorized into:
```
num_ys1 = np.sum(m1[:, : self.n] * m1[:, self.n :], axis=1)
num_ys2 = np.sum(m2[:, : self.n] * m2[:, self.n :], axis=1)
```
I guess it worths doing it.

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 12 above)
Comment (balopat) [reply cirq-core/cirq/qis/clifford_tableau.py:241]: So, I found this, where Theorem 36 describes a vectorized solution: https://arxiv.org/pdf/2009.03218.pdf, it was so small, I implemented it, but the code is not working yet, it fails on some of the random circuit tests, I probably made a mistake somewhere, at least I thoroughly reviewed your tests, and they do seem to be sound :)  

```
       m1 = self.matrix().astype(int)
        m2 = second.matrix().astype(int)

        lmbda = np.zeros((2 * self.n, 2 * self.n))
        lmbda[:self.n, self.n:] = np.eye(self.n)
        p1 = np.diag(m1 @ lmbda @ m1.T)
        m2Lm2T = m2 @ lmbda @ m2.T
        p2 = np.diag(m2Lm2T)

        s1 = self.rs.astype(int)
        s2 = second.rs.astype(int)

        merged_m = np.mod(m1 @ m2, 2)
        p12 = np.mod(np.diag(merged_m @ lmbda @ merged_m.T), 2)
        assert np.allclose(p12, np.mod(p1 + m1 @ p2, 2)), f"expected: {p12}, got: {p1 + m1 @ p2}"

        signs = np.mod(s1 + p1 * (m1 @ p2) + m1 @ s2 + np.diag(
            m1 @ np.tril(np.outer(p2, p2.T) + m2Lm2T, -1) @ m1.T), 2)

        merged_tableau = CliffordTableau(num_qubits=self.n)
        merged_tableau.xs = merged_m[:, :self.n]
        merged_tableau.zs = merged_m[:, self.n:]
        merged_tableau.rs = signs
```

Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py]: Thanks for finding this reference! I spent some time figuring out how to implement it correctly. In coding, it is not as clean as the equations. The ugly part is they use `i^a(-1)^bX^cZ^d` format, where a,b,c,d is a boolean type. We need to convert to this format first, do the computation in that domain, then convert it back. In both conversions, counting the number of y gates is inevitable. Check this modification based on your code:
```
        m1 = self.matrix().astype(int)
        m2 = second.matrix().astype(int)

        num_ys1 = np.sum(m1[:, : self.n] * m1[:, self.n :], axis=1)
        num_ys2 = np.sum(m2[:, : self.n] * m2[:, self.n :], axis=1)

        p1 = np.mod(num_ys1, 2)
        p2 = np.mod(num_ys2, 2)

        s1 = self.rs.astype(int) + np.mod(num_ys1, 4) // 2
        s2 = second.rs.astype(int) + np.mod(num_ys2, 4) // 2

        lmbda = np.zeros((2 * self.n, 2 * self.n))
        lmbda[: self.n, self.n :] = np.eye(self.n)
        m2Lm2T = m2 @ lmbda @ m2.T

        m_12 = np.mod(m1.dot(m2), 2)
        p_12 = np.mod(p1 + m1.dot(p2), 2)
        s_12 = (
            s1
            + m1.dot(s2)
            + p1 * m1.dot(p2)
            + np.diag(m1 @ np.tril(np.outer(p2, p2.T) + m2Lm2T, -1) @ m1.T)
        )
        num_ys12 = np.sum(m_12[:, : self.n] * m_12[:, self.n :], axis=1)
        merged_phase = np.mod(p_12 + 2*s_12 - num_ys12, 4)
        merged_m = m_12

        merged_tableau = CliffordTableau(num_qubits=self.n)
        merged_tableau.xs = merged_m[:, : self.n]
        merged_tableau.zs = merged_m[:, self.n :]
        merged_tableau.rs = merged_phase
```

Comment (balopat) [reply cirq-core/cirq/qis/clifford_tableau.py]: Very nice! I'm impressed how you deeply understood all of this :) 
You are right, the Aaronson & Gottesman representation's `r` is not the same as the `s` vector in the Gosset paper's representation. In order to calculate with the Gosset method, we need to "remove" the "signs" that come from the PauliStrings that have even number of Ys in them. 

I'd be curious to see which implementation is faster. I started doing a bit of benchmarking by modifying the end of your `then` test:

```
    n_qubits = 400
    t_final, t_op, expected_t = _three_identical_table(n_qubits)
    seq_op = random_circuit(num_ops=1000, num_qubits=n_qubits)
    for i, (op, args) in enumerate(seq_op):
            t_op = cirq.CliffordTableau(n_qubits)
            op(t_op, *args)
            t_final = t_final.then(t_op)
            op(expected_t, *args)

    assert expected_t == t_final
```

And started to measure based on n_qubits. 
On my machine 

qubits | new | old
-- | -- | --
100 | 6.56 | 30.8
200 | 61.8 | 105.74
300 | 204.04 | 276.61
400 | 511.63 | 670

Which seems like the first version is slower at first blink, but it seems to be growing slower (with an exponent of ~2 instead of ~3.2)...but we need more datapoints. Both should be O(n^3), so these measurements are probably heavily influenced by vectorization and constant factors. 

So - I like the Gosset version a bit more because we have a nice paper + proof describing it and it looks vectorized, so it should be faster - but we need to confirm it. I think speed does matter here, if anybody will use this for QEC, they will use large tableaux.

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 15]
```diff
@@ -176,6 +179,76 @@ def _str_full_(self) -> str:
...
+        # Any pauli string (one stabilizer) in Clifford Tableau should be able to be expressed as
+        #    (1i)^p (-1)^s X^(mx) Z^(mz)
+        # where p and s are binary scalar and mx and mz are binary vectors.
+        num_ys1 = np.sum(m1[:, : self.n] * m1[:, self.n :], axis=1)
+        num_ys2 = np.sum(m2[:, : self.n] * m2[:, self.n :], axis=1)
+
+        p1 = np.mod(num_ys1, 2)
+        p2 = np.mod(num_ys2, 2)
+
+        s1 = self.rs.astype(int) + np.mod(num_ys1, 4) // 2
```
Comment (balopat) [inline cirq-core/cirq/qis/clifford_tableau.py:222]: Explain the conversion between `s` and `r` (maybe using the YYY example we talked through).

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 16]
```diff
@@ -176,6 +179,76 @@ def _str_full_(self) -> str:
...
+        s2 = second.rs.astype(int) + np.mod(num_ys2, 4) // 2
+
+        lmbda = np.zeros((2 * self.n, 2 * self.n))
+        lmbda[: self.n, self.n :] = np.eye(self.n)
+        m2Lm2T = m2 @ lmbda @ m2.T
```
Comment (balopat) [inline cirq-core/cirq/qis/clifford_tableau.py:227]: we could inline this

**Review by balopat — APPROVED** [review]
LGTM with final nits

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 17]
```diff
@@ -176,6 +179,76 @@ def _str_full_(self) -> str:
...
+
+        m_12 = np.mod(m1.dot(m2), 2)
```
[the remark this answers is not in the document]
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py]: Done.

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 15 above)
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py:225]: Done

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 16 above)
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py]: Done

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 12 above)
Comment (balopat) [reply cirq-core/cirq/qis/clifford_tableau.py]: Just to close this off, the new method is slightly faster but not significantly - my previous data was messed up with manual datataking - this one was automated:

qubits | new | old
-- | -- | --
100 | 24.93 | 30.93
200 | 163.97 | 147.83
300 | 350.11 | 378.99
400 | 807.15 | 928.34
500 | 1679.39 | 1776.07
1000 | 18750.73 | 19171.23
